### PR TITLE
streakcam: acq stream + testcases

### PR DESCRIPTION
**Pull request by Sabrina85**

### CODE REVIEW

File: src/odemis/acq/test/stream_test.py [hunk 1]
```diff
@@ -2200,6 +2201,343 @@ def test_acq_cl_leech(self):
...
+        # update stream (live) to change MCPgain
+        streaks.should_update.value = True
+        # activate/play stream, optical path should be corrected immediately (no need to wait)
+        streaks.is_active.value = True
+
+        # check MCPGain is not 0 as set last when stream gets active
+        self.assertNotEqual(streaks.devMCPgain.value, 0)
+        # checks that HW VA and GUI VA are equal when stream active
+        self.assertEqual(self.streak_unit.MCPgain.value, streaks.devMCPgain.value)
+
```
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/acq/test/stream_test.py]: I tried this now. The .image is a VA of the streakSettingStream? It has so far the correct shape (however transposed). Which MD should the image actually have?

File: src/odemis/acq/test/stream_test.py [hunk 2]
```diff
@@ -2200,6 +2201,343 @@ def test_acq_cl_leech(self):
...
+        sem_width = (emt.shape[0] * emt.pixelSize.value[0],
+                     emt.shape[1] * emt.pixelSize.value[1])
+        # In physical coordinates Y goes up, but in ROI, Y goes down => "1-"
+        pos = (sem_center[0] + sem_width[0] * (roi_center[0] - 0.5),
+               sem_center[1] - sem_width[1] * (roi_center[1] - 0.5))
+
+        logging.debug("Expecting pos %s, pxs %s, res %s", pos, pxs, res)
+        return pos, pxs, res
+
+    def test_streak_playStream(self):
```
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/acq/test/stream_test.py]: I kept the streak in the name similar to ar and arpol. Easier to find stuff I thought. Do you want to have it removed?

File: src/odemis/acq/test/stream_test.py [hunk 3]
```diff
@@ -2200,6 +2201,343 @@ def test_acq_cl_leech(self):
...
+        # the order of setting the HWVAs is TimeRange, StreakMode, MCPGain
+        # MCPGain last as otherwise set to zero due to safety functionality in driver
+        self.assertEqual(streaks.devStreakMode.value, self.streak_unit.streakMode.value)
+        self.assertEqual(streaks.devTimeRange.value, self.streak_unit.timeRange.value)
+        self.assertEqual(streaks.devMCPgain.value, self.streak_unit.MCPgain.value)
+
+        # change VAs --> HW VAs should change as stream is still active
+        streaks.detExposureTime.value = 0.1  # s
+        streaks.detBinning.value = (2, 2)  # TODO check with real HW
+        time.sleep(1)
```
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/acq/test/stream_test.py]: I need to check with the HW again. Think it took some time until the HW changes were correctly applied..

File: src/odemis/acq/stream/_helper.py [hunk 4]
```diff
@@ -505,6 +505,75 @@ def _onNewData(self, dataflow, data):
...
+
+        self.active = False  # variable keep track if stream is active/inactive
+
+        # For SPARC: typical user wants density much lower than SEM
+        self.pixelSize.value *= 30  # TODO still true for streak  What does that mean?
+        self.streak_unit = streak_unit
+        self.streak_delay = streak_delay
+
+        # the VAs are used in SEMCCDMDStream (_sync.py)
+        self._streak_unit_vas = self._duplicateVAs(streak_unit, "dev", streak_unit_vas or set())
```
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/acq/stream/_helper.py]: I tried det. So far it looks okay...

File: src/odemis/acq/stream/_helper.py [hunk 5]
```diff
@@ -505,6 +505,75 @@ def _onNewData(self, dataflow, data):
...
+        # TODO check if we change the timeRange VA in the GUI and a stream is active
+        # also change the triggerDelay via the LUT MD_TIME_RANGE_TO_DELAY
+        # if value not in MD it will use the default value in yaml
+        # maybe report that the delay for that timeRange was not calibrated!
+        # TODO should we have a readonly delay VA in acq tab? might be useful
+        # self.triggerDelay = model.FloatContinuous(triggerDelay, range_trigDelay)
+
+        # whenever .timeRange and/or .streakMode changes
+        # -> set .MCPgain = 0 and update .MCPgain.range
+        self._streak_unit_vas["timeRange"].subscribe(self._OnMCPGain)
```
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/acq/stream/_helper.py]: self._streak_unit_vas = self._duplicateVAs(streak_unit, "det", streak_unit_vas) 
       try:
            self.detTimeRange.subscribe(self._OnMCPGain)
            self.detStreakMode.subscribe(self._OnMCPGain)
            self.detMCPgain.subscribe(self._OnMCPGainRange)
        except ValueError:
            raise ValueError("")

[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/acq/stream/_helper.py]: "protection" = the empty set when duplicating the VAs?

File: src/odemis/acq/test/stream_test.py [hunk 6]
```diff
@@ -2200,6 +2201,356 @@ def test_acq_cl_leech(self):
...
+        # activate/play stream, optical path should be corrected immediately (no need to wait)
+        streaks.is_active.value = True
+
+        time.sleep(2)
+        streaks.is_active.value = False
+
+        self.assertIsNotNone(self._image, "No temporal spectrum received after 2s")
+        self.assertIsInstance(self._image, model.DataArray)
+        # .image should be a 2D temporal spectrum
+        self.assertEqual(self._image.shape[1::-1], streaks.detResolution.value)
```
Comment (pieleric) [inline src/odemis/acq/test/stream_test.py]: Please also check that the .image has the right metadata. It should have the MD_TIME_LIST and MD_WL_LIST.
For now, as it's made to work with the AR canvas, it doesn't matter. However, once we get the proper viewport to work (TemporalSpectrum), then this metadata will be important.

File: src/odemis/acq/test/stream_test.py [hunk 7]
```diff
@@ -2200,6 +2201,356 @@ def test_acq_cl_leech(self):
...
+        tempSpec_da = stss.raw[1]  # temporal spectrum data array
+        shape = tempSpec_da.shape
+        self.assertEqual(shape[3] * shape[4], num_tempSpec)
+        # len of shape should be 5: CTZXY
+        self.assertEqual(len(shape), 5)
+
+        # check if metadata is correctly stored
+        md = tempSpec_da.metadata
+        self.assertIn(model.MD_STREAK_TIMERANGE, md)
+        self.assertIn(model.MD_STREAK_MCPGain, md)
```
Comment (pieleric) [inline src/odemis/acq/test/stream_test.py]: `MD_STREAK_MCPGAIN` (all uppercase). 
Please run the test cases again, to check they still pass with the latest version of the (simulated) driver.

File: src/odemis/acq/test/stream_test.py [hunk 8]
```diff
@@ -2200,6 +2201,356 @@ def test_acq_cl_leech(self):
...
+        self.assertIn(model.MD_POS, md)  # check the corresponding SEM pos is there
+        self.assertIn(model.MD_PIXEL_SIZE, md)  # check the corresponding SEM pos is there
+        self.assertIn(model.MD_WL_LIST, md)
+        self.assertIn(model.MD_TIME_LIST, md)
+
+        md = sem_da.metadata
+        self.assertIn(model.MD_PIXEL_SIZE, md)
+        self.assertIn(model.MD_POS, md)
+
+        # TODO start a new acq and check the estimatedAcq time. something wrong
```
Comment (pieleric) [inline src/odemis/acq/test/stream_test.py]: Let's do this TODO?

File: src/odemis/acq/stream/_helper.py [hunk 9]
```diff
@@ -505,6 +505,71 @@ def _onNewData(self, dataflow, data):
...
+        # This is important for HW safety reasons to not destroy the streak unit,
+        # when changing on of the VA while using a high MCPGain.
+        # While the stream is not active: range of possible values for MCPGain
+        # is limited to values <= current value to also prevent HW damage
+        # when starting to play the stream again.
+        try:
+            self.detStreakMode.subscribe(self._OnStreakSettings)
+            self.detMCPGain.subscribe(self._OnMCPGain)
+        except AttributeError:
+            raise ValueError("Necessary HW VAs for streak camera was not provided")
```
Comment (pieleric) [inline src/odemis/acq/stream/_helper.py]: Please state in the doctstring which VAs are required.

**Pull request merged**